lib/aur-build

```diff
@@ -203,24 +216,42 @@ while true; do
     shift
 done
 
+# mollyguard for makepkg
+if (( UID == 0 )) && { [[ ! -v build_user ]] || [[ ! -v AUR_ASROOT ]]; }; then
```

Comment:
The second condition is always true, instead: `&& ! { [[ -v build_user ]] || [[ -v AUR_ASROOT ]]}` https://en.wikipedia.org/wiki/De_Morgan%27s_laws

Reply:
If both `build_user` and `AUR_ASROOT` are set, the second condition is false, no?

Reply:
`(not A) or (not B)` is always true regardless of truthiness of A or B.

Note that with bash you can use a single `[[` with short-circuiting or/and and even `( expression )` notation (as distinct from subshell).

E.g. If you want to error out if UID is 0 and neither build_user or ABS_ASROOT is set then:

```sh
if [[ $UID == 0 && ! (-v build_user || -v ABS_ASROOT) ]]; then ...
```

As such this will be false if either build_user or ABS_ASROOT is set.

If you want this if statement to be true if both build_user and ABS_ASROOT is set, or if both are unset then you'll need more conditions:

```sh
if [[ $UID == 0 && ( ! (-v build_user || -v ABS_ASROOT) || (-v build_user && -v ABS_ASROOT)) ]]; then
```

With this it will be false if either build_user is set of ABS_ASROOT is set, but not both.

Reply:
> (not A) or (not B) is always true regardless of truthiness of A or B.

I made a truth table:

![img](https://i.imgur.com/F6mmtCB.png)

Demonstration:

```bash
# A true, B true
user_id=0
build_user=
AUR_ASROOT=

(( user_id == 0 )) && { [[ ! -v build_user ]] || [[ ! -v AUR_ASROOT ]]; }
echo $?

# A true, B false
user_id=0
build_user=
unset AUR_ASROOT

(( user_id == 0 )) && { [[ ! -v build_user ]] || [[ ! -v AUR_ASROOT ]]; }
echo $?

# A false, B true
user_id=0
unset build_user
AUR_ASROOT=

(( user_id == 0 )) && { [[ ! -v build_user ]] || [[ ! -v AUR_ASROOT ]]; }
echo $?

# A false, B false
user_id=0
unset build_user
unset AUR_ASROOT

(( user_id == 0 )) && { [[ ! -v build_user ]] || [[ ! -v AUR_ASROOT ]]; }
echo $?
```
```
1
0
0
0
```
So the idea is that the `setpriv` logic requires both `AUR_ASROOT` and `build_user`, otherwise there's an error. With the Morgan, this would be (note the `&&`)
```
(( user_id == 0 )) && ! { [[ -v build_user ]] && [[ -v AUR_ASROOT ]]; }
```

Reply:
Hm, fair enough.  The final construction is nice, but I still think you're better off using `[[` instead of `((` (which enables arbitrary code evaluation in < 5.1 bash).

Reply:
Do you have some more information on `[[`, `((` and code evaulation?

Reply:
http://mywiki.wooledge.org/BashProgramming/05#Arithmetic_Expansion

It also applies to `test -v`, see also https://lists.gnu.org/archive/html/bug-bash/2021-10/msg00175.html

Reply:
Thanks, both changes are in 9.2 now.